students/cfarris/lessons/lesson10/assignment/databse.py

```diff
@@ -0,0 +1,252 @@
+#!/usr/bin/env Python3
+
+"""
+This set up is only for development use. Not production
+Uses Pymongo to access MongoDB database.
+Imports 3 files, directs them to different collections
+Retrieves information necessary for HP Norton collection
+
+lesson10 Database.py. Measure performance.
+
+"""
+import csv
+from pymongo import MongoClient
+import time
+import types
+
+
+
+class Timer:  
+    def __init__(self, func=time.perf_counter):
+        self.elapsed = 0.0
+        self._func = func
+        self._start = None
+
+    def start(self):
+        if self._start is not None:
+            raise RuntimeError('Already started')
+        self._start = self._func()
+
+    def stop(self):
+        if self._start is None:
+            raise RuntimeError('Not started')
+        end = self._func()
+        self.elapsed += end - self._start
+        self._start = None
+
+    def reset(self):
+        self.elapsed = 0.0
+
+    @property
+    def running(self):
+        return self._start is not None
+
+    def __enter__(self):
+        self.start()
+        return self
+
+    def __exit__(self, *args):
+        self.stop()
+
+
+def timefunc(fn, *args, **kwargs):
+
+    def fncomposite(*args, **kwargs):
+        timer = Timer()
+        timer.start()
+        rt = fn(*args, **kwargs)
+        timer.stop()
+        #print("Executing %s took %s seconds." % (fn.__name__, timer.elapsed))
+        eval_statement = ("Executing %s took %s seconds." % (fn.__name__, timer.elapsed))
+        #print(eval_statement)
+        timing_list.append(eval_statement)
+        return rt
+    # return the composite function
+    return fncomposite
+
+
+class MetaTimer(type): #Timer Class
+    #
+    def __new__(cls, clsname, bases, attr):
+        #else execute __new__method of super class
+        #cal __init__ of type class
+        for name, value in attr.items():
+            if type(value) is types.FunctionType or type(value) is types.MethodType:
+                attr[name] = timefunc(value)
+
+        return super(MetaTimer, cls).__new__(cls, clsname, bases, attr) #standard fare when you are doing metaclasses.
+
+
+class MongoDBConnection(metaclass=MetaTimer):
+    """Connect to MongoDB"""
+    def __init__(self, host='127.0.0.1', port=27017):
+
+        self.host = host
+        self.port = port
+        self.connection = None
+
+
+    def __enter__(self):
+        self.connection = MongoClient(self.host, self.port)
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.connection.close()
```

Comment:
I don't think it is very useful to time the connection to the mongo database, but yes, this is how you would do it.